text/0000-get-collection-keys.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: get_collection_keys
+- Start Date: 2015-06-24
+- RFC PR: https://github.com/rust-lang/rust/pull/26531
+- Rust Issue:
+
+# Summary
+
+Get the original keys back from collections.
+
+# Motivation
+
+Removing a key from a HashMap or an item from a HashSet will drop the item.
+This is usually acceptable for small clonable objects like numbers and strings,
+but they are not always clonable, and even if they are it might be an expensive
+operation.
+
+# Detailed design
+
+There are currently four operations that could have a new function to them:
+`get`, `get_mut`, `insert` and `remove`. All of those methods receive a reference
+to a key type and return reference to the value or a the value itself.
+
+New analogous methods can be added to return not just the value but both the key
+and the value. For `insert` and `remove` they can return the key and the value,
+while `get` and `get_mut` will return a reference to them. Notice that `get_mut`
+will return a mutable value but not a mutable key.
+
+In the case of HashSet, the same concept applies for the items as if they were
+keys in a hash map. It does not make sense to implement something like `get_mut`
+in this case.
+
+Structs that should implement these new methods are:
+
+* std::collections::BTreeMap
+* std::collections::HashMap
+* std::collections::BTreeSet
+* std::collections::HashSet
+
+
+Some structs might implement these new methods, but their benefit besides
+consistency is unclear since it would only apply for `usize`:
+
+* std::collections::VecMap
+* std::collections::BitSet
+
+New API methods:
+
+```rust
+impl<K: Ord, V> BTreeMap<K, V> {
+    // ...
+
+    pub fn keyed_get<Q: ?Sized>(&self, key: &Q) -> Option<(&K, &V)> where K: Borrow<Q>, Q: Ord;
+    pub fn keyed_get_mut<Q: ?Sized>(&mut self, key: &Q) -> Option<(&K, &mut V)> where K: Borrow<Q>, Q: Ord;
```

Comment:
Arguably you can yield `&mut K` for the hypothetical desire to mutate things in the key that don't affect hash/eq/ord. This of course allows you to mess up hash/eq/ord but the collection can't possibly care since that can already be true thanks to:

* Interior mutability
* Bad implementations of hash/eq/ord

The collection is already locked up yielding the `&mut V`, so it's not an ownership problem. In fact the `&K` is *also* mutably borrowing the map because of how lifetimes work. Obviously mutating a key in place is *super sketchy*, and I'm not clear that anyone can reasonably suggest they want this. Still... you could.